File: go/0347_top_k_frequent_elements.go
package main

func topKFrequent(nums []int, k int) (res []int) {
	countMap := map[int]int{}
	for _, num := range nums {
		if count, ok := countMap[num]; ok {
			countMap[num] = count + 1
		} else {
			countMap[num] = 1
		}
	}

	bucket := make([][]int, len(nums)+1)
	for num, count := range countMap {
		bucket[count] = append(bucket[count], num)
	}

	for i := len(bucket) - 1; i > 0; i-- {
		res = append(res, bucket[i]...)
		if len(res) == k {
			return
		}
	}

	return
}

File: csharp/0347-top-k-frequent-elements.cs
public class Solution {
    public int[] TopKFrequent(int[] nums, int k) {
        var count = new Dictionary<int, int>();
        var freq = new List<int>[nums.Length + 1];

        for (int i = 0; i < freq.Length; i++) {
            freq[i] = new List<int>();
        }

        foreach (int n in nums) {
            if (count.ContainsKey(n)) {
                count[n]++;
            } else {
                count[n] = 1;
            }
        }

        foreach (var pair in count) {
            freq[pair.Value].Add(pair.Key);
        }

        var res = new List<int>();
        for (int i = freq.Length - 1; i > 0; i--) {
            foreach (int n in freq[i]) {
                res.Add(n);
                if (res.Count == k) {
                    return res.ToArray();
                }
            }
        }

        return res.ToArray();
    }